// @gofynd/fp-signature/dist/cjs/RequestSigner.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const querystring_1 = __importDefault(require("querystring"));
const constants_1 = require("./constants");
const utils_1 = require("./utils");
let ALL_HEADERS_TO_INCLUDE = [...constants_1.HEADERS_TO_INCLUDE];
class RequestSigner {
    constructor(request, secret, headersToInclude) {
        if (!secret) {
            throw new Error("Signature secrete cannot be null, pass secret parameter in constructor.");
        }
        if (headersToInclude && !Array.isArray(headersToInclude)) {
            throw new Error("Headers to include must be an array");
        }
        this.secret = secret;
        this.request = request;
        ALL_HEADERS_TO_INCLUDE = [...ALL_HEADERS_TO_INCLUDE, ...(headersToInclude || [])];
        let headers = (this.request.headers = this.request.headers || {});
        if (!this.request.method && this.request.body) {
            this.request.method = "POST";
        }
        if (!headers.Host && !headers.host) {
            headers.Host = this.request.host;
            // If a port is specified explicitly, use it as is
            if (this.request.port) {
                headers.Host += ":" + this.request.port;
            }
        }
        if (!this.request.host) {
            this.request.host = headers.Host || headers.host;
        }
    }
    prepareRequest() {
        this.parsePath();
        let request = this.request;
        let headers = request.headers;
        let query;
        if (request.signQuery) {
            this.parsedPath.query = query = this.parsedPath.query || {};
            if (query["x-fp-date"]) {
                this.datetime = query["x-fp-date"];
            }
            else {
                query["x-fp-date"] = this.getDateTime();
            }
        }
        else {
            if (!request.doNotModifyHeaders) {
                if (headers["x-fp-date"]) {
                    this.datetime = headers["x-fp-date"] || headers["x-fp-date"];
                }
                else {
                    headers["x-fp-date"] = this.getDateTime();
                }
            }
            delete headers["x-fp-signature"];
            delete headers["X-Fp-Signature"];
        }
    }
    sign() {
        this.request.signQuery = false;
        if (!this.parsedPath) {
            this.prepareRequest();
        }
        this.request.headers["x-fp-signature"] = this.signature();
        return {
            'x-fp-signature': this.request.headers['x-fp-signature'],
            'x-fp-date': this.request.headers["x-fp-date"]
        };
    }
    signQuery() {
        this.request.signQuery = true;
        if (!this.parsedPath) {
            this.prepareRequest();
        }
        this.parsedPath.query["x-fp-signature"] = this.signature();
        this.request.path = this.formatPath();
        return {
            'x-fp-signature': this.parsedPath.query['x-fp-signature'],
            'x-fp-date': this.parsedPath.query["x-fp-date"]
        };
    }
    getDateTime() {
        if (!this.datetime) {
            let headers = this.request.headers;
            let date = new Date(headers.Date || headers.date || new Date());
            this.datetime = date.toISOString().replace(/[:\-]|\.\d{3}/g, "");
        }
        return this.datetime;
    }
    signature() {
        let strTosign = this.stringToSign();
        return `v1.1:${(0, utils_1.hmac)(this.secret, strTosign, "hex")}`;
    }
    stringToSign() {
        return [this.getDateTime(), (0, utils_1.hash)(this.canonicalString(), "hex")].join("\n");
    }
    canonicalString() {
        if (!this.parsedPath) {
            this.prepareRequest();
        }
        let pathStr = this.parsedPath.path;
        let query = this.parsedPath.query;
        let headers = this.request.headers;
        let queryStr = "";
        let normalizePath = true;
        let decodePath = this.request.doNotEncodePath;
        let decodeSlashesInPath = false;
        let firstValOnly = false;
        let bodyHash = (0, utils_1.hash)(this.request.body || "", "hex");
        if (query) {
            let reducedQuery = Object.keys(query).reduce(function (obj, key) {
                var _a;
                if (!key) {
                    return obj;
                }
                obj[(0, utils_1.encodeRfc3986Full)(key)] = !Array.isArray(query[key])
                    ? query[key]
                    : firstValOnly
                        ? (_a = query[key]) === null || _a === void 0 ? void 0 : _a[0]
                        : query[key];
                return obj;
            }, {});
            let encodedQueryPieces = [];
            Object.keys(reducedQuery)
                .sort()
                .forEach(function (key) {
                if (!Array.isArray(reducedQuery[key])) {
                    encodedQueryPieces.push(key + "=" + (0, utils_1.encodeRfc3986Full)(reducedQuery[key]));
                }
                else {
                    reducedQuery[key]
                        .map(utils_1.encodeRfc3986Full)
                        .sort()
                        .forEach(function (val) {
                        encodedQueryPieces.push(key + "=" + val);
                    });
                }
            });
            queryStr = encodedQueryPieces.join("&");
        }
        if (pathStr !== "/") {
            if (normalizePath) {
                pathStr = pathStr.replace(/\/{2,}/g, "/");
            }
            pathStr = pathStr
                .split("/")
                .reduce(function (path, piece) {
                if (normalizePath && piece === "..") {
                    path.pop();
                }
                else if (!normalizePath || piece !== ".") {
                    if (decodePath)
                        piece = decodeURIComponent(piece.replace(/\+/g, " "));
                    path.push((0, utils_1.encodeRfc3986Full)(piece));
                }
                return path;
            }, [])
                .join("/");
            if (pathStr[0] !== "/")
                pathStr = "/" + pathStr;
            if (decodeSlashesInPath)
                pathStr = pathStr.replace(/%2F/g, "/");
        }
        let canonicalReq = [
            this.request.method || "GET",
            pathStr,
            queryStr,
            this.canonicalHeaders() + "\n",
            this.signedHeaders(),
            bodyHash,
        ].join("\n");
        return canonicalReq;
    }
    canonicalHeaders() {
        let headers = this.request.headers;
        function trimAll(header) {
            return header.toString().trim().replace(/\s+/g, " ");
        }
        return Object.keys(headers)
            .filter(function (key) {
            let notInIgnoreHeader = constants_1.HEADERS_TO_IGNORE[key.toLowerCase()] == null;
            if (notInIgnoreHeader) {
                let foundMatch = false;
                for (let t in ALL_HEADERS_TO_INCLUDE) {
                    foundMatch =
                        foundMatch || new RegExp(ALL_HEADERS_TO_INCLUDE[t], "ig").test(key);
                }
                return foundMatch;
            }
            else {
                return false;
            }
        })
            .sort(function (a, b) {
            return a.toLowerCase() < b.toLowerCase() ? -1 : 1;
        })
            .map(function (key) {
            return key.toLowerCase() + ":" + trimAll(headers[key]);
        })
            .join("\n");
    }
    signedHeaders() {
        return Object.keys(this.request.headers)
            .map(function (key) {
            return key.toLowerCase();
        })
            .filter(function (key) {
            let notInIgnoreHeader = constants_1.HEADERS_TO_IGNORE[key.toLowerCase()] == null;
            if (notInIgnoreHeader) {
                let foundMatch = false;
                for (let t in ALL_HEADERS_TO_INCLUDE) {
                    foundMatch =
                        foundMatch || new RegExp(ALL_HEADERS_TO_INCLUDE[t], "ig").test(key);
                }
                return foundMatch;
            }
            else {
                return false;
            }
        })
            .sort()
            .join(";");
    }
    parsePath() {
        let path = this.request.path || "/";
        let queryIx = path.indexOf("?");
        let query = {};
        if (queryIx >= 0) {
            query = querystring_1.default.parse(path.slice(queryIx + 1));
            delete query["x-fp-signature"];
            delete query["X-Fp-Signature"];
            path = path.slice(0, queryIx);
        }
        path = path
            .split("/")
            .map((t) => {
            return encodeURIComponent(decodeURIComponent(t));
        })
            .join("/");
        this.parsedPath = {
            path: path,
            query: query,
        };
    }
    formatPath() {
        let path = this.parsedPath.path;
        let query = this.parsedPath.query;
        if (!query) {
            return path;
        }
        // Services don't support empty query string keys
        if (query[""] != null) {
            delete query[""];
        }
        return path + "?" + (0, utils_1.encodeRfc3986)(querystring_1.default.stringify(query));
    }
}
exports.default = RequestSigner;
